xarray/core/treenode.py

```diff
@@ -681,7 +691,8 @@ def __str__(self) -> str:
 
     def _post_attach(self: AnyNamedNode, parent: AnyNamedNode, name: str) -> None:
         """Ensures child has name attribute corresponding to key under which it has been stored."""
-        self.name = name
+        _validate_name(name)  # is this check redundant?
+        self._name = name
```

Comment:
Yes it's redundant, at least in the sense that the original code uses the `.name` property setter, which now calls `_validate_name`. Other than that it is necessary though, because you can get down to this method by following 

https://github.com/pydata/xarray/blob/18e5c87696c2233202bc67066408e6f495d0327d/xarray/core/treenode.py#L180

Reply:
Actually you could instead move this check to be inside the `._check_children` staticmethod, which will get called first. That might be a clearer place to put the check.

Reply:
As an aside I think this brings us closer to seeing that `NamedNode` is almost entirely unnecessary, if it weren't for the possibility of a root node having a name. Using `TreeNode` alone nodes would still have names, but those names are  stored only as child keys on the parent. 

So it makes sense that here we are able to refactor such that every name validation is actually done on `TreeNode`, with the exceptions of inside `NamedNode.__init__` and the `.name` setter, which are the only places we could possibly check the name of the root node (as it has no parent to perform the validation via).

Reply:
hmm was I wrong about this?

Reply:
it does seems so... honestly I find the various hooks on the base TreeNode pretty hard to reason about

Reply:
Damn. I guess just put the validation check back in, merge, then we can refactor with that regression test to adhere to?

> I find the various hooks on the base TreeNode pretty hard to reason about

I normally find them okay - I want to look at this issue more closely. Of all the hooks defined on TreeNode we only actually use like 2 of them, the rest are holdovers from the original design copying the anytree library.